{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-12-20 09:42:43.027220: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-12-20 09:42:43.042670: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-12-20 09:42:43.057677: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-12-20 09:42:43.061962: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-12-20 09:42:43.076186: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-12-20 09:42:45.690321: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import scanpy.external as sce\n",
    "import scirpy as ir\n",
    "import decoupler as dc\n",
    "# import scrublet as scr\n",
    "import muon as mu\n",
    "#from vpolo.alevin import parser # to parse alevin output\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.stats.multitest as multi\n",
    "from adjustText import adjust_text\n",
    "import re\n",
    "\n",
    "seed = 0\n",
    "np.random.seed(seed)\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['figure.facecolor'] = (1,1,1,1)\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.10.2 anndata==0.10.9 umap==0.5.6 numpy==1.26.4 scipy==1.11.4 pandas==2.0.3 scikit-learn==1.5.1 statsmodels==0.14.2 igraph==0.11.6 pynndescent==0.5.13\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80, color_map='viridis', transparent=False, frameon=False)  # low dpi (dots per inch) yields small inline figures\n",
    "\n",
    "import matplotlib as mpl\n",
    "# 2 lines below solved the facecolor problem.\n",
    "# mpl.rcParams['figure.facecolor'] = 'white'\n",
    "mpl.rcParams['figure.facecolor'] = (1,1,1,1)\n",
    "sc.settings.autosave = True\n",
    "sc.logging.print_header()\n",
    "\n",
    "version = '241219_startrac'\n",
    "# input_table = '../data/231009_PBMC_HH514.csv'\n",
    "\n",
    "results_file = '../scanpy/{}/res.h5ad'.format(version)\n",
    "results_file_mu = '../scanpy/{}/res.h5mu'.format(version)\n",
    "results_file_cellxgene = '../scanpy/{}/res.cxg.h5ad'.format(version)\n",
    "\n",
    "import os\n",
    "os.makedirs('../scanpy/{}'.format(version), exist_ok=True)\n",
    "\n",
    "sc.settings.figdir = '../scanpy/{}/graph'.format(version)\n",
    "sc.settings.cachedir = '../scanpy/{}/cache'.format(version)\n",
    "# %config InlineBackend.figure_format = 'retina'\n",
    "\n",
    "import os\n",
    "os.makedirs('../scanpy/{}'.format(version), exist_ok=True)\n",
    "os.makedirs(sc.settings.figdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_csf = sc.read('/home/yy693/pi_hafler/ASAP/scanpy/241028_CSF_downstream/res.h5ad')\n",
    "adata_bld = sc.read('/home/yy693/pi_hafler/ASAP/scanpy/241028_BLD_downstream/res.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.concat([adata_bld, adata_csf])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellranger_sample_id</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>DonorID</th>\n",
       "      <th>Estimated Number of Cells</th>\n",
       "      <th>Mean Reads per Cell</th>\n",
       "      <th>Median Genes per Cell</th>\n",
       "      <th>Number of Reads</th>\n",
       "      <th>Valid Barcodes</th>\n",
       "      <th>Sequencing Saturation</th>\n",
       "      <th>...</th>\n",
       "      <th>majority_voting_Immune_All_Low</th>\n",
       "      <th>predicted_labels_COVID19_HumanChallenge_Blood</th>\n",
       "      <th>over_clustering_COVID19_HumanChallenge_Blood</th>\n",
       "      <th>majority_voting_COVID19_HumanChallenge_Blood</th>\n",
       "      <th>cluster_L1</th>\n",
       "      <th>cluster_L2</th>\n",
       "      <th>Hyposmia</th>\n",
       "      <th>Diagnosis_RBDstrat</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>T CD8 EMRA</td>\n",
       "      <td>291</td>\n",
       "      <td>T CD8 EMRA</td>\n",
       "      <td>T</td>\n",
       "      <td>CD8CD4 Temra</td>\n",
       "      <td>False</td>\n",
       "      <td>HC</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>Classical monocytes</td>\n",
       "      <td>Monocyte CD14+</td>\n",
       "      <td>284</td>\n",
       "      <td>Monocyte CD14+</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>CD14 Mono</td>\n",
       "      <td>False</td>\n",
       "      <td>HC</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCCCTTG-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>CD16+ NK cells</td>\n",
       "      <td>NK</td>\n",
       "      <td>57</td>\n",
       "      <td>NK</td>\n",
       "      <td>T</td>\n",
       "      <td>CD8CD4 Temra</td>\n",
       "      <td>False</td>\n",
       "      <td>HC</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCCCACG-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tem/Trm cytotoxic T cells</td>\n",
       "      <td>T CD8 EM</td>\n",
       "      <td>218</td>\n",
       "      <td>T CD8 EM</td>\n",
       "      <td>T</td>\n",
       "      <td>CD8 Tem</td>\n",
       "      <td>False</td>\n",
       "      <td>HC</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACGAGAGT-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>T CD4 CTL</td>\n",
       "      <td>104</td>\n",
       "      <td>T CD8 CTL</td>\n",
       "      <td>T</td>\n",
       "      <td>CD8CD4 Temra</td>\n",
       "      <td>False</td>\n",
       "      <td>HC</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGTTTATC-1-YPD028CSF</th>\n",
       "      <td>YPD028CSF_HHT_GEX</td>\n",
       "      <td>YPD028CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>14,782</td>\n",
       "      <td>24,075</td>\n",
       "      <td>1,586</td>\n",
       "      <td>355,879,272</td>\n",
       "      <td>88.1%</td>\n",
       "      <td>71.3%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>446</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>TNK</td>\n",
       "      <td>CD4 Tem</td>\n",
       "      <td>True</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTCGCATC-1-YPD028CSF</th>\n",
       "      <td>YPD028CSF_HHT_GEX</td>\n",
       "      <td>YPD028CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>14,782</td>\n",
       "      <td>24,075</td>\n",
       "      <td>1,586</td>\n",
       "      <td>355,879,272</td>\n",
       "      <td>88.1%</td>\n",
       "      <td>71.3%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>T CD4 Helper 1</td>\n",
       "      <td>552</td>\n",
       "      <td>T CD4 Helper 1</td>\n",
       "      <td>TNK</td>\n",
       "      <td>CD4 Tem</td>\n",
       "      <td>True</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGATTCT-1-YPD028CSF</th>\n",
       "      <td>YPD028CSF_HHT_GEX</td>\n",
       "      <td>YPD028CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>14,782</td>\n",
       "      <td>24,075</td>\n",
       "      <td>1,586</td>\n",
       "      <td>355,879,272</td>\n",
       "      <td>88.1%</td>\n",
       "      <td>71.3%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>T CD4 Helper 2</td>\n",
       "      <td>94</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>TNK</td>\n",
       "      <td>CD4 Tcm</td>\n",
       "      <td>True</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGGTGTA-1-YPD028CSF</th>\n",
       "      <td>YPD028CSF_HHT_GEX</td>\n",
       "      <td>YPD028CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>14,782</td>\n",
       "      <td>24,075</td>\n",
       "      <td>1,586</td>\n",
       "      <td>355,879,272</td>\n",
       "      <td>88.1%</td>\n",
       "      <td>71.3%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>94</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>TNK</td>\n",
       "      <td>CD4 Tcm</td>\n",
       "      <td>True</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTAACCT-1-YPD028CSF</th>\n",
       "      <td>YPD028CSF_HHT_GEX</td>\n",
       "      <td>YPD028CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>14,782</td>\n",
       "      <td>24,075</td>\n",
       "      <td>1,586</td>\n",
       "      <td>355,879,272</td>\n",
       "      <td>88.1%</td>\n",
       "      <td>71.3%</td>\n",
       "      <td>...</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>99</td>\n",
       "      <td>T CD4 Helper 0</td>\n",
       "      <td>TNK</td>\n",
       "      <td>CD4 Tcm</td>\n",
       "      <td>True</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>783081 rows × 257 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             cellranger_sample_id   SampleID SampleType  \\\n",
       "index                                                                     \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD    YPD001BLD_HHT_GEX  YPD001BLD        BLD   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD    YPD001BLD_HHT_GEX  YPD001BLD        BLD   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD    YPD001BLD_HHT_GEX  YPD001BLD        BLD   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD    YPD001BLD_HHT_GEX  YPD001BLD        BLD   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD    YPD001BLD_HHT_GEX  YPD001BLD        BLD   \n",
       "...                                           ...        ...        ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF    YPD028CSF_HHT_GEX  YPD028CSF        CSF   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF    YPD028CSF_HHT_GEX  YPD028CSF        CSF   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF    YPD028CSF_HHT_GEX  YPD028CSF        CSF   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF    YPD028CSF_HHT_GEX  YPD028CSF        CSF   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF    YPD028CSF_HHT_GEX  YPD028CSF        CSF   \n",
       "\n",
       "                             DonorID Estimated Number of Cells  \\\n",
       "index                                                            \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD  YPD001                     6,931   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD  YPD001                     6,931   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD  YPD001                     6,931   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD  YPD001                     6,931   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD  YPD001                     6,931   \n",
       "...                              ...                       ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF  YPD028                    14,782   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF  YPD028                    14,782   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF  YPD028                    14,782   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF  YPD028                    14,782   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF  YPD028                    14,782   \n",
       "\n",
       "                             Mean Reads per Cell Median Genes per Cell  \\\n",
       "index                                                                    \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD              47,589                 2,450   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD              47,589                 2,450   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD              47,589                 2,450   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD              47,589                 2,450   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD              47,589                 2,450   \n",
       "...                                          ...                   ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF              24,075                 1,586   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF              24,075                 1,586   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF              24,075                 1,586   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF              24,075                 1,586   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF              24,075                 1,586   \n",
       "\n",
       "                             Number of Reads Valid Barcodes  \\\n",
       "index                                                         \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD     329,839,529          90.1%   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD     329,839,529          90.1%   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD     329,839,529          90.1%   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD     329,839,529          90.1%   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD     329,839,529          90.1%   \n",
       "...                                      ...            ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF     355,879,272          88.1%   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF     355,879,272          88.1%   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF     355,879,272          88.1%   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF     355,879,272          88.1%   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF     355,879,272          88.1%   \n",
       "\n",
       "                             Sequencing Saturation  ...  \\\n",
       "index                                               ...   \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                 75.5%  ...   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                 75.5%  ...   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                 75.5%  ...   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                 75.5%  ...   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                 75.5%  ...   \n",
       "...                                            ...  ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF                 71.3%  ...   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF                 71.3%  ...   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF                 71.3%  ...   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF                 71.3%  ...   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF                 71.3%  ...   \n",
       "\n",
       "                             majority_voting_Immune_All_Low  \\\n",
       "index                                                         \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD    Tem/Temra cytotoxic T cells   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD            Classical monocytes   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                 CD16+ NK cells   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD      Tem/Trm cytotoxic T cells   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD    Tem/Temra cytotoxic T cells   \n",
       "...                                                     ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF       Tcm/Naive helper T cells   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF       Tcm/Naive helper T cells   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF       Tcm/Naive helper T cells   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF       Tcm/Naive helper T cells   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF       Tcm/Naive helper T cells   \n",
       "\n",
       "                             predicted_labels_COVID19_HumanChallenge_Blood  \\\n",
       "index                                                                        \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                                    T CD8 EMRA   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                                Monocyte CD14+   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                                            NK   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                                      T CD8 EM   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                                     T CD4 CTL   \n",
       "...                                                                    ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF                                T CD4 Helper 0   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF                                T CD4 Helper 1   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF                                T CD4 Helper 2   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF                                T CD4 Helper 0   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF                                T CD4 Helper 0   \n",
       "\n",
       "                             over_clustering_COVID19_HumanChallenge_Blood  \\\n",
       "index                                                                       \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                                          291   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                                          284   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                                           57   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                                          218   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                                          104   \n",
       "...                                                                   ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF                                          446   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF                                          552   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF                                           94   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF                                           94   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF                                           99   \n",
       "\n",
       "                             majority_voting_COVID19_HumanChallenge_Blood  \\\n",
       "index                                                                       \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                                   T CD8 EMRA   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                               Monocyte CD14+   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                                           NK   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                                     T CD8 EM   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                                    T CD8 CTL   \n",
       "...                                                                   ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF                               T CD4 Helper 0   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF                               T CD4 Helper 1   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF                               T CD4 Helper 0   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF                               T CD4 Helper 0   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF                               T CD4 Helper 0   \n",
       "\n",
       "                             cluster_L1    cluster_L2 Hyposmia  \\\n",
       "index                                                            \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD          T  CD8CD4 Temra    False   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD    Myeloid     CD14 Mono    False   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD          T  CD8CD4 Temra    False   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD          T       CD8 Tem    False   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD          T  CD8CD4 Temra    False   \n",
       "...                                 ...           ...      ...   \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF        TNK       CD4 Tem     True   \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF        TNK       CD4 Tem     True   \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF        TNK       CD4 Tcm     True   \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF        TNK       CD4 Tcm     True   \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF        TNK       CD4 Tcm     True   \n",
       "\n",
       "                             Diagnosis_RBDstrat  Gender   Age  \n",
       "index                                                          \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                 HC    Male  79.0  \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                 HC    Male  79.0  \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                 HC    Male  79.0  \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                 HC    Male  79.0  \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                 HC    Male  79.0  \n",
       "...                                         ...     ...   ...  \n",
       "TTTGTCATCGTTTATC-1-YPD028CSF             PD-RBD  Female  77.0  \n",
       "TTTGTCATCTCGCATC-1-YPD028CSF             PD-RBD  Female  77.0  \n",
       "TTTGTCATCTGATTCT-1-YPD028CSF             PD-RBD  Female  77.0  \n",
       "TTTGTCATCTGGTGTA-1-YPD028CSF             PD-RBD  Female  77.0  \n",
       "TTTGTCATCTTAACCT-1-YPD028CSF             PD-RBD  Female  77.0  \n",
       "\n",
       "[783081 rows x 257 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_csf\n",
    "del adata_bld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/utils.py:346: ExperimentalFeatureWarning: Support for Awkward Arrays is currently experimental. Behavior may change in the future. Please report any issues you may encounter!\n",
      "  warnings.warn(msg, category, stacklevel=stacklevel)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: Non-standard locus name: None \n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/utils.py:346: ExperimentalFeatureWarning: Outer joins on awkward.Arrays will have different return values in the future. For details, and to offer input, please see:\n",
      "\n",
      "\thttps://github.com/scverse/anndata/issues/898\n",
      "  warnings.warn(msg, category, stacklevel=stacklevel)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: Non-standard locus name: None \n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: Non-standard locus name: None \n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "WARNING: `locus` column not found in input data. The locus is being inferred from the {v,d,j,c}_call columns.\n",
      "Filtering chains...\n",
      "Indexing VJ chains...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing VDJ chains...\n",
      "build result array\n",
      "Stored result in `adata.obs[\"receptor_type\"]`.\n",
      "Stored result in `adata.obs[\"receptor_subtype\"]`.\n",
      "Stored result in `adata.obs[\"chain_pairing\"]`.\n",
      "Filtering chains...\n",
      "Indexing VJ chains...\n",
      "Indexing VDJ chains...\n",
      "build result array\n",
      "Stored result in `adata.obs[\"receptor_type\"]`.\n",
      "Stored result in `adata.obs[\"receptor_subtype\"]`.\n",
      "Stored result in `adata.obs[\"chain_pairing\"]`.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 1029222 × 23891\n",
       "  3 modalities\n",
       "    gex:\t783081 x 23891\n",
       "      obs:\t&#x27;cellranger_sample_id&#x27;, &#x27;SampleID&#x27;, &#x27;SampleType&#x27;, &#x27;DonorID&#x27;, &#x27;Estimated Number of Cells&#x27;, &#x27;Mean Reads per Cell&#x27;, &#x27;Median Genes per Cell&#x27;, &#x27;Number of Reads&#x27;, &#x27;Valid Barcodes&#x27;, &#x27;Sequencing Saturation&#x27;, &#x27;Q30 Bases in Barcode&#x27;, &#x27;Q30 Bases in RNA Read&#x27;, &#x27;Q30 Bases in UMI&#x27;, &#x27;Reads Mapped to Genome&#x27;, &#x27;Reads Mapped Confidently to Genome&#x27;, &#x27;Reads Mapped Confidently to Intergenic Regions&#x27;, &#x27;Reads Mapped Confidently to Intronic Regions&#x27;, &#x27;Reads Mapped Confidently to Exonic Regions&#x27;, &#x27;Reads Mapped Confidently to Transcriptome&#x27;, &#x27;Reads Mapped Antisense to Gene&#x27;, &#x27;Fraction Reads in Cells&#x27;, &#x27;Total Genes Detected&#x27;, &#x27;Median UMI Counts per Cell&#x27;, &#x27;Q30 Bases in RNA Read 2&#x27;, &#x27;chemistry&#x27;, &#x27;Visit&#x27;, &#x27;SubjectID&#x27;, &#x27;DateICFSigned&#x27;, &#x27;YearICFSigned&#x27;, &#x27;EnrollmentCategory&#x27;, &#x27;RBDY/N?&#x27;, &#x27;YearofRBDDx&#x27;, &#x27;RBDAAO&#x27;, &#x27;AgeRBDDx&#x27;, &#x27;RBDDxtoConesnt&#x27;, &#x27;PDY/N&#x27;, &#x27;Year1stPDSx&#x27;, &#x27;YearofPDDx&#x27;, &#x27;PDAAO&#x27;, &#x27;AgePDDx&#x27;, &#x27;PDDxtoConsent&#x27;, &#x27;DLBY/N&#x27;, &#x27;Year1stDLBSx&#x27;, &#x27;YearDLBDX&#x27;, &#x27;DLBAAO&#x27;, &#x27;AgeDLBDx&#x27;, &#x27;DLBDxtoConsent&#x27;, &#x27;PSGinEpic?&#x27;, &#x27;BirthYear&#x27;, &#x27;MeetsMDSPDCriteria?&#x27;, &#x27;PSGDate&#x27;, &#x27;Ethnicity&#x27;, &#x27;Race&#x27;, &#x27;RaceOther&#x27;, &#x27;PDFamilyHx&#x27;, &#x27;PDFHMother&#x27;, &#x27;PDFHFather&#x27;, &#x27;PDFHFullSibs&#x27;, &#x27;PDFHHalfSibs&#x27;, &#x27;PDFHMaternalGPs&#x27;, &#x27;PDFHPaternalGPs&#x27;, &#x27;PDFHMaternalAuntsorUncles&#x27;, &#x27;PDFHChildren&#x27;, &#x27;PDFHOther&#x27;, &#x27;DLBFamilyHx&#x27;, &#x27;DLBFHMother&#x27;, &#x27;DLBFHFather&#x27;, &#x27;DLBFHFullSibs&#x27;, &#x27;DLBFHHalfSibs&#x27;, &#x27;DLBFHMaternalGPs&#x27;, &#x27;DLBFHPaternalGPs&#x27;, &#x27;DLBFHMaternalAuntsorUncles&#x27;, &#x27;DLBFHChildren&#x27;, &#x27;DLBFHOther&#x27;, &#x27;MSAFamilyHx&#x27;, &#x27;MSAFHMother&#x27;, &#x27;MSAFHFather&#x27;, &#x27;MSAFHFullSibs&#x27;, &#x27;MSAFHHalfSibs&#x27;, &#x27;MSAFHMaternalGPs&#x27;, &#x27;MSAFHPaternalGPs&#x27;, &#x27;MSAFHMaternalAuntsorUncles&#x27;, &#x27;MSAFHChildren&#x27;, &#x27;MSAFHOther&#x27;, &#x27;PDRiskPesticideExposure&#x27;, &#x27;PDRiskOccupationalsolventExposure&#x27;, &#x27;PDRiskDrinksCaffeinatedBeverages&#x27;, &#x27;PDRiskSmokerorFormerSmoker&#x27;, &#x27;PDRiskPhysicallyActive&#x27;, &#x27;MeanStriatumSBR&#x27;, &#x27;MeanStriatumSBR%ile&#x27;, &#x27;Caudate/PutamenRatio&#x27;, &#x27;AssymetryIndex&#x27;, &#x27;OrthostaticHypotension?&#x27;, &#x27;BloodyTap?&#x27;, &#x27;ClinicalCSFDataObtained?&#x27;, &#x27;CSFProtein&#x27;, &#x27;CSFGlucose&#x27;, &#x27;ClinicalCSFRBC&#x27;, &#x27;ClinicalCSFTotalNucleatedCells&#x27;, &#x27;ClinicalCSFGranulocytes&#x27;, &#x27;ClinicalCSFLymphocytes&#x27;, &#x27;CSFClinicalMonocytes&#x27;, &#x27;CSFVolume&#x27;, &#x27;CSFTotalCells&#x27;, &#x27;CSFCells/uL&#x27;, &#x27;CSFNotes&#x27;, &#x27;CSFSAA&#x27;, &#x27;UPSITRawScore&#x27;, &#x27;UPSITPercentile&#x27;, &#x27;MOCAVisuospatial&#x27;, &#x27;MOCA2Naming&#x27;, &#x27;MOCAMemoryTrial1&#x27;, &#x27;MOCAMemoryTrial2&#x27;, &#x27;MOCAAttention&#x27;, &#x27;MOCALanguage&#x27;, &#x27;MOCAAbstraction&#x27;, &#x27;MOCADelayedRecall&#x27;, &#x27;MOCAMemoryIndexScore&#x27;, &#x27;MOCAOrientation&#x27;, &#x27;MOCATotal&#x27;, &#x27;ParisArousalTotal&#x27;, &#x27;ConstipationCriteriaScore&#x27;, &#x27;BMSQ1RisingfromChair&#x27;, &#x27;BMSQ2Handwriting&#x27;, &#x27;BMSQ3VoiceSofter&#x27;, &#x27;BMSQ4PoorBalance&#x27;, &#x27;BMSQ5FeetGettingStuck&#x27;, &#x27;BMSQ6FaceLessExpressive&#x27;, &#x27;BMSQ7ArmsorLegsShaking&#x27;, &#x27;BMSQ8Buttons&#x27;, &#x27;BMSQ9Shuffling&#x27;, &#x27;BMSQ10MoveMoreSlowlythanOthersofAge&#x27;, &#x27;BMSQTotal&#x27;, &#x27;LRScore&#x27;, &#x27;ProbableProdromalPD?&#x27;, &#x27;SCOPATotal&#x27;, &#x27;SCPOPAManualEntry&#x27;, &#x27;NMSSTotal&#x27;, &#x27;MDS-UPDRS1.1cognition&#x27;, &#x27;MDS-UPDRS1.2Hallucinations&#x27;, &#x27;MDS-UPDRS1.3Mood&#x27;, &#x27;MDS-UPDRS1.4Anxiety&#x27;, &#x27;MDS-UPDRS1.5Apathy&#x27;, &#x27;MDS-UPDRS1.6DOPADysregulation&#x27;, &#x27;MDS-UPDRS1.7Sleep&#x27;, &#x27;MDS-UPDRS1.8DaytimeSleepiness&#x27;, &#x27;MDS-UPDRS1.9Pain&#x27;, &#x27;UMD-UPDRS1.10Urological&#x27;, &#x27;MDS-UPDRS1.11Constipation&#x27;, &#x27;MDS-UPDRS1.12OrthostaticDizziness&#x27;, &#x27;MDS-UPDRS1.13Fatigue&#x27;, &#x27;MDS-UPDRS2.1Speech&#x27;, &#x27;MDS-UPDRS2.2Drooling&#x27;, &#x27;MDS-UPDRS2.3Swallowing&#x27;, &#x27;MDS-UPDRS2.4Eating&#x27;, &#x27;MDS-UPDRS2.5Dressing&#x27;, &#x27;MDS-UPDRS2.6Hygeine&#x27;, &#x27;MDS-UPDRS2.7Handwriting&#x27;, &#x27;MDS-UPDRS2.8Hobbies&#x27;, &#x27;MDS-UPDRS2.9TurninginBed&#x27;, &#x27;MDS-UPDRS2.10Tremor&#x27;, &#x27;MDS-UPDRS2.11RisingfromChair,Car,Bed&#x27;, &#x27;MDS-UPDRS2.12Walking&#x27;, &#x27;MDS-UPDRS2.13FreezingofGait&#x27;, &#x27;MDS-UPDRSTakingMeds&#x27;, &#x27;MDS-UPDRS3.1Speech&#x27;, &#x27;MDS-UPDRS3.2Facies&#x27;, &#x27;MDS-UPDRS3.3NeckRigidity&#x27;, &#x27;MDS-UPDRS3.3RUERigidity&#x27;, &#x27;MDS-UPDRS3.3LUERidigity&#x27;, &#x27;MDS-UPDRS3.3RLERidigity&#x27;, &#x27;MDS-UPDRS3.3LLERigidty&#x27;, &#x27;MDS-UPDRS3.4FingerTapsR&#x27;, &#x27;MDS-UPDRS3.4FingerTapsL&#x27;, &#x27;MDS-UPDRS3.5GraspR&#x27;, &#x27;MDS-UPDRS3.5GraspL&#x27;, &#x27;MDS-UPDRS3.6RSMR&#x27;, &#x27;MDS-UPDRS3.6RSML&#x27;, &#x27;MDS-UPDRS3.7ToeTapR&#x27;, &#x27;MDS-UPDRS3.7ToeTapL&#x27;, &#x27;MDS-UPDRS3.8LegAgilityR&#x27;, &#x27;MDS-UPDRS3.8LegAgilityL&#x27;, &#x27;MDS-UPDRS3.9RisingFromChair&#x27;, &#x27;MDS-UPDRS3.10Gait&#x27;, &#x27;MDS-UPDRS3.11FreezingofGait&#x27;, &#x27;MDS-UPDRS3.12PullTest&#x27;, &#x27;MDS-UPDRS3.13Posture&#x27;, &#x27;MDS-UPDRS3.14GlobalSpontenaity&#x27;, &#x27;MDS-UPDRS3.15PosturalTremorR&#x27;, &#x27;MDS-UPDRS3.15PosturalTremorL&#x27;, &#x27;MDS=UPDRS3.16KineticTremorRUE&#x27;, &#x27;MDS=UPDRS3.16KineticTremorLUE&#x27;, &#x27;MDS=UPDRS3.17RestTremorAmp.RUE&#x27;, &#x27;MDS=UPDRS3.17RestTremorAmp.LUE&#x27;, &#x27;MDS=UPDRS3.17RestTremorAmp.RLE&#x27;, &#x27;MDS=UPDRS3.17RestTremorAmp.LLE&#x27;, &#x27;MDS=UPDRS3.17RestTremorAmp.Lips/Jaw&#x27;, &#x27;MDS-UPDRS3.18ConstancyofRestTremor&#x27;, &#x27;Hoehn&amp;YahrStage&#x27;, &#x27;MDS-UPDRSPartITotal&#x27;, &#x27;MDS-UPDRSPartIITotal&#x27;, &#x27;MDS-UPDRSPartIIITotal&#x27;, &#x27;MDS-UPDRSPartI-IIITotal&#x27;, &#x27;MDS-UPDRSPartIbTotal&#x27;, &#x27;MDS-UPDRSPartIb+IITotal&#x27;, &#x27;RBDSQTotal&#x27;, &#x27;ESSTotal&#x27;, &#x27;RightCaudateSBR&#x27;, &#x27;RightAnteriorPutamenSBR&#x27;, &#x27;RightPosteiorPutamenSBR&#x27;, &#x27;LeftAnteirorPutamenSBR&#x27;, &#x27;LeftPosteriorPutamenSBR&#x27;, &#x27;LeftCaudateSBR&#x27;, &#x27;MeanStriatumSBR.1&#x27;, &#x27;MeanStriatumSBRPercentile&#x27;, &#x27;CaudatePutamenRatio&#x27;, &#x27;AssymetryIndex.1&#x27;, &#x27;PhenoconverteasofSept2024r?&#x27;, &#x27;CSFTNFalpha&#x27;, &#x27;CSFTNFBeta&#x27;, &#x27;PlasmaTNFalpha&#x27;, &#x27;PlasmaTNFBeta&#x27;, &#x27;doublet_score&#x27;, &#x27;predicted_doublet&#x27;, &#x27;n_genes_by_counts&#x27;, &#x27;log1p_n_genes_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;pct_counts_in_top_50_genes&#x27;, &#x27;pct_counts_in_top_100_genes&#x27;, &#x27;pct_counts_in_top_200_genes&#x27;, &#x27;pct_counts_in_top_500_genes&#x27;, &#x27;total_counts_mt&#x27;, &#x27;log1p_total_counts_mt&#x27;, &#x27;pct_counts_mt&#x27;, &#x27;total_counts_ribo&#x27;, &#x27;log1p_total_counts_ribo&#x27;, &#x27;pct_counts_ribo&#x27;, &#x27;total_counts_hb&#x27;, &#x27;log1p_total_counts_hb&#x27;, &#x27;pct_counts_hb&#x27;, &#x27;n_genes&#x27;, &#x27;_scvi_batch&#x27;, &#x27;_scvi_labels&#x27;, &#x27;predicted_labels_Immune_All_Low&#x27;, &#x27;over_clustering_Immune_All_Low&#x27;, &#x27;majority_voting_Immune_All_Low&#x27;, &#x27;predicted_labels_COVID19_HumanChallenge_Blood&#x27;, &#x27;over_clustering_COVID19_HumanChallenge_Blood&#x27;, &#x27;majority_voting_COVID19_HumanChallenge_Blood&#x27;, &#x27;cluster_L1&#x27;, &#x27;cluster_L2&#x27;, &#x27;Hyposmia&#x27;, &#x27;Diagnosis_RBDstrat&#x27;, &#x27;Gender&#x27;, &#x27;Age&#x27;\n",
       "      obsm:\t&#x27;X_scVI&#x27;, &#x27;X_umap&#x27;, &#x27;X_umap_filtered&#x27;, &#x27;_scvi_extra_categorical_covs&#x27;, &#x27;_scvi_extra_continuous_covs&#x27;\n",
       "      layers:\t&#x27;counts&#x27;\n",
       "    tcr:\t638242 x 0\n",
       "      obs:\t&#x27;cellranger_sample_id&#x27;, &#x27;SampleID&#x27;, &#x27;SampleType&#x27;, &#x27;DonorID&#x27;, &#x27;receptor_type&#x27;, &#x27;receptor_subtype&#x27;, &#x27;chain_pairing&#x27;\n",
       "      uns:\t&#x27;chain_indices&#x27;\n",
       "      obsm:\t&#x27;airr&#x27;, &#x27;chain_indices&#x27;\n",
       "    bcr:\t69214 x 0\n",
       "      obs:\t&#x27;cellranger_sample_id&#x27;, &#x27;SampleID&#x27;, &#x27;SampleType&#x27;, &#x27;DonorID&#x27;, &#x27;receptor_type&#x27;, &#x27;receptor_subtype&#x27;, &#x27;chain_pairing&#x27;\n",
       "      uns:\t&#x27;chain_indices&#x27;\n",
       "      obsm:\t&#x27;airr&#x27;, &#x27;chain_indices&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 1029222 × 23891\n",
       "  3 modalities\n",
       "    gex:\t783081 x 23891\n",
       "      obs:\t'cellranger_sample_id', 'SampleID', 'SampleType', 'DonorID', 'Estimated Number of Cells', 'Mean Reads per Cell', 'Median Genes per Cell', 'Number of Reads', 'Valid Barcodes', 'Sequencing Saturation', 'Q30 Bases in Barcode', 'Q30 Bases in RNA Read', 'Q30 Bases in UMI', 'Reads Mapped to Genome', 'Reads Mapped Confidently to Genome', 'Reads Mapped Confidently to Intergenic Regions', 'Reads Mapped Confidently to Intronic Regions', 'Reads Mapped Confidently to Exonic Regions', 'Reads Mapped Confidently to Transcriptome', 'Reads Mapped Antisense to Gene', 'Fraction Reads in Cells', 'Total Genes Detected', 'Median UMI Counts per Cell', 'Q30 Bases in RNA Read 2', 'chemistry', 'Visit', 'SubjectID', 'DateICFSigned', 'YearICFSigned', 'EnrollmentCategory', 'RBDY/N?', 'YearofRBDDx', 'RBDAAO', 'AgeRBDDx', 'RBDDxtoConesnt', 'PDY/N', 'Year1stPDSx', 'YearofPDDx', 'PDAAO', 'AgePDDx', 'PDDxtoConsent', 'DLBY/N', 'Year1stDLBSx', 'YearDLBDX', 'DLBAAO', 'AgeDLBDx', 'DLBDxtoConsent', 'PSGinEpic?', 'BirthYear', 'MeetsMDSPDCriteria?', 'PSGDate', 'Ethnicity', 'Race', 'RaceOther', 'PDFamilyHx', 'PDFHMother', 'PDFHFather', 'PDFHFullSibs', 'PDFHHalfSibs', 'PDFHMaternalGPs', 'PDFHPaternalGPs', 'PDFHMaternalAuntsorUncles', 'PDFHChildren', 'PDFHOther', 'DLBFamilyHx', 'DLBFHMother', 'DLBFHFather', 'DLBFHFullSibs', 'DLBFHHalfSibs', 'DLBFHMaternalGPs', 'DLBFHPaternalGPs', 'DLBFHMaternalAuntsorUncles', 'DLBFHChildren', 'DLBFHOther', 'MSAFamilyHx', 'MSAFHMother', 'MSAFHFather', 'MSAFHFullSibs', 'MSAFHHalfSibs', 'MSAFHMaternalGPs', 'MSAFHPaternalGPs', 'MSAFHMaternalAuntsorUncles', 'MSAFHChildren', 'MSAFHOther', 'PDRiskPesticideExposure', 'PDRiskOccupationalsolventExposure', 'PDRiskDrinksCaffeinatedBeverages', 'PDRiskSmokerorFormerSmoker', 'PDRiskPhysicallyActive', 'MeanStriatumSBR', 'MeanStriatumSBR%ile', 'Caudate/PutamenRatio', 'AssymetryIndex', 'OrthostaticHypotension?', 'BloodyTap?', 'ClinicalCSFDataObtained?', 'CSFProtein', 'CSFGlucose', 'ClinicalCSFRBC', 'ClinicalCSFTotalNucleatedCells', 'ClinicalCSFGranulocytes', 'ClinicalCSFLymphocytes', 'CSFClinicalMonocytes', 'CSFVolume', 'CSFTotalCells', 'CSFCells/uL', 'CSFNotes', 'CSFSAA', 'UPSITRawScore', 'UPSITPercentile', 'MOCAVisuospatial', 'MOCA2Naming', 'MOCAMemoryTrial1', 'MOCAMemoryTrial2', 'MOCAAttention', 'MOCALanguage', 'MOCAAbstraction', 'MOCADelayedRecall', 'MOCAMemoryIndexScore', 'MOCAOrientation', 'MOCATotal', 'ParisArousalTotal', 'ConstipationCriteriaScore', 'BMSQ1RisingfromChair', 'BMSQ2Handwriting', 'BMSQ3VoiceSofter', 'BMSQ4PoorBalance', 'BMSQ5FeetGettingStuck', 'BMSQ6FaceLessExpressive', 'BMSQ7ArmsorLegsShaking', 'BMSQ8Buttons', 'BMSQ9Shuffling', 'BMSQ10MoveMoreSlowlythanOthersofAge', 'BMSQTotal', 'LRScore', 'ProbableProdromalPD?', 'SCOPATotal', 'SCPOPAManualEntry', 'NMSSTotal', 'MDS-UPDRS1.1cognition', 'MDS-UPDRS1.2Hallucinations', 'MDS-UPDRS1.3Mood', 'MDS-UPDRS1.4Anxiety', 'MDS-UPDRS1.5Apathy', 'MDS-UPDRS1.6DOPADysregulation', 'MDS-UPDRS1.7Sleep', 'MDS-UPDRS1.8DaytimeSleepiness', 'MDS-UPDRS1.9Pain', 'UMD-UPDRS1.10Urological', 'MDS-UPDRS1.11Constipation', 'MDS-UPDRS1.12OrthostaticDizziness', 'MDS-UPDRS1.13Fatigue', 'MDS-UPDRS2.1Speech', 'MDS-UPDRS2.2Drooling', 'MDS-UPDRS2.3Swallowing', 'MDS-UPDRS2.4Eating', 'MDS-UPDRS2.5Dressing', 'MDS-UPDRS2.6Hygeine', 'MDS-UPDRS2.7Handwriting', 'MDS-UPDRS2.8Hobbies', 'MDS-UPDRS2.9TurninginBed', 'MDS-UPDRS2.10Tremor', 'MDS-UPDRS2.11RisingfromChair,Car,Bed', 'MDS-UPDRS2.12Walking', 'MDS-UPDRS2.13FreezingofGait', 'MDS-UPDRSTakingMeds', 'MDS-UPDRS3.1Speech', 'MDS-UPDRS3.2Facies', 'MDS-UPDRS3.3NeckRigidity', 'MDS-UPDRS3.3RUERigidity', 'MDS-UPDRS3.3LUERidigity', 'MDS-UPDRS3.3RLERidigity', 'MDS-UPDRS3.3LLERigidty', 'MDS-UPDRS3.4FingerTapsR', 'MDS-UPDRS3.4FingerTapsL', 'MDS-UPDRS3.5GraspR', 'MDS-UPDRS3.5GraspL', 'MDS-UPDRS3.6RSMR', 'MDS-UPDRS3.6RSML', 'MDS-UPDRS3.7ToeTapR', 'MDS-UPDRS3.7ToeTapL', 'MDS-UPDRS3.8LegAgilityR', 'MDS-UPDRS3.8LegAgilityL', 'MDS-UPDRS3.9RisingFromChair', 'MDS-UPDRS3.10Gait', 'MDS-UPDRS3.11FreezingofGait', 'MDS-UPDRS3.12PullTest', 'MDS-UPDRS3.13Posture', 'MDS-UPDRS3.14GlobalSpontenaity', 'MDS-UPDRS3.15PosturalTremorR', 'MDS-UPDRS3.15PosturalTremorL', 'MDS=UPDRS3.16KineticTremorRUE', 'MDS=UPDRS3.16KineticTremorLUE', 'MDS=UPDRS3.17RestTremorAmp.RUE', 'MDS=UPDRS3.17RestTremorAmp.LUE', 'MDS=UPDRS3.17RestTremorAmp.RLE', 'MDS=UPDRS3.17RestTremorAmp.LLE', 'MDS=UPDRS3.17RestTremorAmp.Lips/Jaw', 'MDS-UPDRS3.18ConstancyofRestTremor', 'Hoehn&YahrStage', 'MDS-UPDRSPartITotal', 'MDS-UPDRSPartIITotal', 'MDS-UPDRSPartIIITotal', 'MDS-UPDRSPartI-IIITotal', 'MDS-UPDRSPartIbTotal', 'MDS-UPDRSPartIb+IITotal', 'RBDSQTotal', 'ESSTotal', 'RightCaudateSBR', 'RightAnteriorPutamenSBR', 'RightPosteiorPutamenSBR', 'LeftAnteirorPutamenSBR', 'LeftPosteriorPutamenSBR', 'LeftCaudateSBR', 'MeanStriatumSBR.1', 'MeanStriatumSBRPercentile', 'CaudatePutamenRatio', 'AssymetryIndex.1', 'PhenoconverteasofSept2024r?', 'CSFTNFalpha', 'CSFTNFBeta', 'PlasmaTNFalpha', 'PlasmaTNFBeta', 'doublet_score', 'predicted_doublet', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', '_scvi_batch', '_scvi_labels', 'predicted_labels_Immune_All_Low', 'over_clustering_Immune_All_Low', 'majority_voting_Immune_All_Low', 'predicted_labels_COVID19_HumanChallenge_Blood', 'over_clustering_COVID19_HumanChallenge_Blood', 'majority_voting_COVID19_HumanChallenge_Blood', 'cluster_L1', 'cluster_L2', 'Hyposmia', 'Diagnosis_RBDstrat', 'Gender', 'Age'\n",
       "      obsm:\t'X_scVI', 'X_umap', 'X_umap_filtered', '_scvi_extra_categorical_covs', '_scvi_extra_continuous_covs'\n",
       "      layers:\t'counts'\n",
       "    tcr:\t638242 x 0\n",
       "      obs:\t'cellranger_sample_id', 'SampleID', 'SampleType', 'DonorID', 'receptor_type', 'receptor_subtype', 'chain_pairing'\n",
       "      uns:\t'chain_indices'\n",
       "      obsm:\t'airr', 'chain_indices'\n",
       "    bcr:\t69214 x 0\n",
       "      obs:\t'cellranger_sample_id', 'SampleID', 'SampleType', 'DonorID', 'receptor_type', 'receptor_subtype', 'chain_pairing'\n",
       "      uns:\t'chain_indices'\n",
       "      obsm:\t'airr', 'chain_indices'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_file_list = '/home/yy693/pi_hafler/ASAP/data/241219_cellranger_files_VDJ.csv'\n",
    "\n",
    "\n",
    "df_file = pd.read_csv(f_file_list)\n",
    "\n",
    "list_cols = ['cellranger_sample_id', 'SampleID', \n",
    "       'SampleType', 'DonorID']\n",
    "       \n",
    "dict_adata_tcr = {}\n",
    "for pos,row in df_file[df_file['vdj_type']=='TCR'].iterrows():\n",
    "    a = ir.io.read_airr(f\"{row['path']}/airr_rearrangement.tsv\")\n",
    "    a.var_names_make_unique()\n",
    "    for col in list_cols:\n",
    "        a.obs[col] = row[col]\n",
    "\n",
    "    # sc.pp.scrublet(a, expected_doublet_rate=doublet_rate(a.shape[0], row['chemistry']))\n",
    "\n",
    "    dict_adata_tcr[row['SampleID']] = a\n",
    "\n",
    "adata_tcr = sc.concat(dict_adata_tcr, join='outer', index_unique='-')\n",
    "\n",
    "dict_adata_bcr = {}\n",
    "for pos,row in df_file[df_file['vdj_type']=='BCR'].iterrows():\n",
    "    if int(row['Estimated Number of Cells'].replace(',','')) > 0:\n",
    "        a = ir.io.read_airr(f\"{row['path']}/airr_rearrangement.tsv\")\n",
    "        a.var_names_make_unique()\n",
    "        for col in list_cols:\n",
    "            a.obs[col] = row[col]\n",
    "\n",
    "        # sc.pp.scrublet(a, expected_doublet_rate=doublet_rate(a.shape[0], row['chemistry']))\n",
    "\n",
    "        dict_adata_bcr[row['SampleID']] = a\n",
    "\n",
    "adata_bcr = sc.concat(dict_adata_bcr, join='outer', index_unique='-')\n",
    "\n",
    "ir.pp.index_chains(adata_tcr)\n",
    "ir.tl.chain_qc(adata_tcr)\n",
    "adata_tcr = adata_tcr[adata_tcr.obs['receptor_type'] == 'TCR']\n",
    "\n",
    "ir.pp.index_chains(adata_bcr)\n",
    "ir.tl.chain_qc(adata_bcr)\n",
    "adata_bcr = adata_bcr[adata_bcr.obs['receptor_type'] == 'BCR']\n",
    "\n",
    "mdata = mu.MuData({\"gex\": adata, \"tcr\": adata_tcr, \"bcr\": adata_bcr})\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gex:cellranger_sample_id</th>\n",
       "      <th>gex:SampleID</th>\n",
       "      <th>gex:SampleType</th>\n",
       "      <th>gex:DonorID</th>\n",
       "      <th>gex:Estimated Number of Cells</th>\n",
       "      <th>gex:Mean Reads per Cell</th>\n",
       "      <th>gex:Median Genes per Cell</th>\n",
       "      <th>gex:Number of Reads</th>\n",
       "      <th>gex:Valid Barcodes</th>\n",
       "      <th>gex:Sequencing Saturation</th>\n",
       "      <th>...</th>\n",
       "      <th>tcr:receptor_type</th>\n",
       "      <th>tcr:receptor_subtype</th>\n",
       "      <th>tcr:chain_pairing</th>\n",
       "      <th>bcr:cellranger_sample_id</th>\n",
       "      <th>bcr:SampleID</th>\n",
       "      <th>bcr:SampleType</th>\n",
       "      <th>bcr:DonorID</th>\n",
       "      <th>bcr:receptor_type</th>\n",
       "      <th>bcr:receptor_subtype</th>\n",
       "      <th>bcr:chain_pairing</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCCCTTG-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCCCACG-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>TCR</td>\n",
       "      <td>TRA+TRB</td>\n",
       "      <td>extra VJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACGAGAGT-1-YPD001BLD</th>\n",
       "      <td>YPD001BLD_HHT_GEX</td>\n",
       "      <td>YPD001BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>6,931</td>\n",
       "      <td>47,589</td>\n",
       "      <td>2,450</td>\n",
       "      <td>329,839,529</td>\n",
       "      <td>90.1%</td>\n",
       "      <td>75.5%</td>\n",
       "      <td>...</td>\n",
       "      <td>TCR</td>\n",
       "      <td>TRA+TRB</td>\n",
       "      <td>single pair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCTTCGGTTCCACAA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YPD094BLD_BCR_VHT_VDJ</td>\n",
       "      <td>YPD094BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD094</td>\n",
       "      <td>BCR</td>\n",
       "      <td>IGH+IGK</td>\n",
       "      <td>single pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTTCTAAGCCCGAAA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YPD094BLD_BCR_VHT_VDJ</td>\n",
       "      <td>YPD094BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD094</td>\n",
       "      <td>BCR</td>\n",
       "      <td>IGH+IGK</td>\n",
       "      <td>single pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGAGGCTCTTGCATT-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YPD094BLD_BCR_VHT_VDJ</td>\n",
       "      <td>YPD094BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD094</td>\n",
       "      <td>BCR</td>\n",
       "      <td>IGH+IGL</td>\n",
       "      <td>single pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCTCCTTCGTTACGA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YPD094BLD_BCR_VHT_VDJ</td>\n",
       "      <td>YPD094BLD</td>\n",
       "      <td>BLD</td>\n",
       "      <td>YPD094</td>\n",
       "      <td>BCR</td>\n",
       "      <td>IGH+IGK</td>\n",
       "      <td>single pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGACGTCTCGCAAACT-1-YPD094CSF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YPD094CSF_BCR_VHT_VDJ</td>\n",
       "      <td>YPD094CSF</td>\n",
       "      <td>CSF</td>\n",
       "      <td>YPD094</td>\n",
       "      <td>BCR</td>\n",
       "      <td>IGH+IGK</td>\n",
       "      <td>single pair</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1029222 rows × 271 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gex:cellranger_sample_id gex:SampleID  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD        YPD001BLD_HHT_GEX    YPD001BLD   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD        YPD001BLD_HHT_GEX    YPD001BLD   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD        YPD001BLD_HHT_GEX    YPD001BLD   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD        YPD001BLD_HHT_GEX    YPD001BLD   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD        YPD001BLD_HHT_GEX    YPD001BLD   \n",
       "...                                               ...          ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                      NaN          NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                      NaN          NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                      NaN          NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                      NaN          NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                      NaN          NaN   \n",
       "\n",
       "                             gex:SampleType gex:DonorID  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD            BLD      YPD001   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD            BLD      YPD001   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD            BLD      YPD001   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD            BLD      YPD001   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD            BLD      YPD001   \n",
       "...                                     ...         ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD            NaN         NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD            NaN         NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD            NaN         NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD            NaN         NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF            NaN         NaN   \n",
       "\n",
       "                             gex:Estimated Number of Cells  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                         6,931   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                         6,931   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                         6,931   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                         6,931   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                         6,931   \n",
       "...                                                    ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                           NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                           NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                           NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                           NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                           NaN   \n",
       "\n",
       "                             gex:Mean Reads per Cell  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                  47,589   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                  47,589   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                  47,589   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                  47,589   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                  47,589   \n",
       "...                                              ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                     NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                     NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                     NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                     NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                     NaN   \n",
       "\n",
       "                             gex:Median Genes per Cell gex:Number of Reads  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                     2,450         329,839,529   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD                     2,450         329,839,529   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                     2,450         329,839,529   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                     2,450         329,839,529   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                     2,450         329,839,529   \n",
       "...                                                ...                 ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                       NaN                 NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                       NaN                 NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                       NaN                 NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                       NaN                 NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                       NaN                 NaN   \n",
       "\n",
       "                             gex:Valid Barcodes gex:Sequencing Saturation  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD              90.1%                     75.5%   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD              90.1%                     75.5%   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD              90.1%                     75.5%   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD              90.1%                     75.5%   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD              90.1%                     75.5%   \n",
       "...                                         ...                       ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                NaN                       NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                NaN                       NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                NaN                       NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                NaN                       NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                NaN                       NaN   \n",
       "\n",
       "                              ... tcr:receptor_type tcr:receptor_subtype  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD  ...               NaN                  NaN   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD  ...               NaN                  NaN   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD  ...               NaN                  NaN   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD  ...               TCR              TRA+TRB   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD  ...               TCR              TRA+TRB   \n",
       "...                           ...               ...                  ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD  ...               NaN                  NaN   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD  ...               NaN                  NaN   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD  ...               NaN                  NaN   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD  ...               NaN                  NaN   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF  ...               NaN                  NaN   \n",
       "\n",
       "                             tcr:chain_pairing bcr:cellranger_sample_id  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD               NaN                      NaN   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD               NaN                      NaN   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD               NaN                      NaN   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD          extra VJ                      NaN   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD       single pair                      NaN   \n",
       "...                                        ...                      ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD               NaN    YPD094BLD_BCR_VHT_VDJ   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD               NaN    YPD094BLD_BCR_VHT_VDJ   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD               NaN    YPD094BLD_BCR_VHT_VDJ   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD               NaN    YPD094BLD_BCR_VHT_VDJ   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF               NaN    YPD094CSF_BCR_VHT_VDJ   \n",
       "\n",
       "                             bcr:SampleID bcr:SampleType bcr:DonorID  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD          NaN            NaN         NaN   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD          NaN            NaN         NaN   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD          NaN            NaN         NaN   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD          NaN            NaN         NaN   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD          NaN            NaN         NaN   \n",
       "...                                   ...            ...         ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD    YPD094BLD            BLD      YPD094   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD    YPD094BLD            BLD      YPD094   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD    YPD094BLD            BLD      YPD094   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD    YPD094BLD            BLD      YPD094   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF    YPD094CSF            CSF      YPD094   \n",
       "\n",
       "                             bcr:receptor_type bcr:receptor_subtype  \\\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD               NaN                  NaN   \n",
       "AAACCTGAGGACATTA-1-YPD001BLD               NaN                  NaN   \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD               NaN                  NaN   \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD               NaN                  NaN   \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD               NaN                  NaN   \n",
       "...                                        ...                  ...   \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD               BCR              IGH+IGK   \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD               BCR              IGH+IGK   \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD               BCR              IGH+IGL   \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD               BCR              IGH+IGK   \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF               BCR              IGH+IGK   \n",
       "\n",
       "                             bcr:chain_pairing  \n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD               NaN  \n",
       "AAACCTGAGGACATTA-1-YPD001BLD               NaN  \n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD               NaN  \n",
       "AAACCTGAGTCCCACG-1-YPD001BLD               NaN  \n",
       "AAACCTGCACGAGAGT-1-YPD001BLD               NaN  \n",
       "...                                        ...  \n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD       single pair  \n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD       single pair  \n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD       single pair  \n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD       single pair  \n",
       "GGACGTCTCGCAAACT-1-YPD094CSF       single pair  \n",
       "\n",
       "[1029222 rows x 271 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing sequence x sequence distance matrix for VJ sequences.\n",
      "Computing sequence x sequence distance matrix for VDJ sequences.\n",
      "Initializing lookup tables. \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/scirpy/ir_dist/__init__.py:252: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.\n",
      "  params.adata.uns[key_added] = result\n",
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/gpfs/gibbs/project/hafler/yy693/conda_envs/scanpy1.10.2/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--> Done initializing lookup tables. (0:00:51)\n",
      "Computing clonotype x clonotype distances.\n",
      "--> Done computing clonotype x clonotype distances.  (0:00:02)\n",
      "Stored result in `mdata.obs[\"tcr:clone_id\"]`.\n",
      "Stored result in `mdata.obs[\"tcr:clone_id_size\"]`.\n"
     ]
    }
   ],
   "source": [
    "# using default parameters, `ir_dist` will compute nucleotide sequence identity\n",
    "ir.pp.ir_dist(mdata, airr_mod='tcr')\n",
    "ir.tl.define_clonotypes(mdata, airr_mod='tcr', receptor_arms=\"all\", dual_ir=\"primary_only\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gex:cellranger_sample_id',\n",
       " 'gex:SampleID',\n",
       " 'gex:SampleType',\n",
       " 'gex:DonorID',\n",
       " 'gex:Estimated Number of Cells',\n",
       " 'gex:Mean Reads per Cell',\n",
       " 'gex:Median Genes per Cell',\n",
       " 'gex:Number of Reads',\n",
       " 'gex:Valid Barcodes',\n",
       " 'gex:Sequencing Saturation',\n",
       " 'gex:Q30 Bases in Barcode',\n",
       " 'gex:Q30 Bases in RNA Read',\n",
       " 'gex:Q30 Bases in UMI',\n",
       " 'gex:Reads Mapped to Genome',\n",
       " 'gex:Reads Mapped Confidently to Genome',\n",
       " 'gex:Reads Mapped Confidently to Intergenic Regions',\n",
       " 'gex:Reads Mapped Confidently to Intronic Regions',\n",
       " 'gex:Reads Mapped Confidently to Exonic Regions',\n",
       " 'gex:Reads Mapped Confidently to Transcriptome',\n",
       " 'gex:Reads Mapped Antisense to Gene',\n",
       " 'gex:Fraction Reads in Cells',\n",
       " 'gex:Total Genes Detected',\n",
       " 'gex:Median UMI Counts per Cell',\n",
       " 'gex:Q30 Bases in RNA Read 2',\n",
       " 'gex:chemistry',\n",
       " 'gex:Visit',\n",
       " 'gex:SubjectID',\n",
       " 'gex:DateICFSigned',\n",
       " 'gex:YearICFSigned',\n",
       " 'gex:EnrollmentCategory',\n",
       " 'gex:RBDY/N?',\n",
       " 'gex:YearofRBDDx',\n",
       " 'gex:RBDAAO',\n",
       " 'gex:AgeRBDDx',\n",
       " 'gex:RBDDxtoConesnt',\n",
       " 'gex:PDY/N',\n",
       " 'gex:Year1stPDSx',\n",
       " 'gex:YearofPDDx',\n",
       " 'gex:PDAAO',\n",
       " 'gex:AgePDDx',\n",
       " 'gex:PDDxtoConsent',\n",
       " 'gex:DLBY/N',\n",
       " 'gex:Year1stDLBSx',\n",
       " 'gex:YearDLBDX',\n",
       " 'gex:DLBAAO',\n",
       " 'gex:AgeDLBDx',\n",
       " 'gex:DLBDxtoConsent',\n",
       " 'gex:PSGinEpic?',\n",
       " 'gex:BirthYear',\n",
       " 'gex:MeetsMDSPDCriteria?',\n",
       " 'gex:PSGDate',\n",
       " 'gex:Ethnicity',\n",
       " 'gex:Race',\n",
       " 'gex:RaceOther',\n",
       " 'gex:PDFamilyHx',\n",
       " 'gex:PDFHMother',\n",
       " 'gex:PDFHFather',\n",
       " 'gex:PDFHFullSibs',\n",
       " 'gex:PDFHHalfSibs',\n",
       " 'gex:PDFHMaternalGPs',\n",
       " 'gex:PDFHPaternalGPs',\n",
       " 'gex:PDFHMaternalAuntsorUncles',\n",
       " 'gex:PDFHChildren',\n",
       " 'gex:PDFHOther',\n",
       " 'gex:DLBFamilyHx',\n",
       " 'gex:DLBFHMother',\n",
       " 'gex:DLBFHFather',\n",
       " 'gex:DLBFHFullSibs',\n",
       " 'gex:DLBFHHalfSibs',\n",
       " 'gex:DLBFHMaternalGPs',\n",
       " 'gex:DLBFHPaternalGPs',\n",
       " 'gex:DLBFHMaternalAuntsorUncles',\n",
       " 'gex:DLBFHChildren',\n",
       " 'gex:DLBFHOther',\n",
       " 'gex:MSAFamilyHx',\n",
       " 'gex:MSAFHMother',\n",
       " 'gex:MSAFHFather',\n",
       " 'gex:MSAFHFullSibs',\n",
       " 'gex:MSAFHHalfSibs',\n",
       " 'gex:MSAFHMaternalGPs',\n",
       " 'gex:MSAFHPaternalGPs',\n",
       " 'gex:MSAFHMaternalAuntsorUncles',\n",
       " 'gex:MSAFHChildren',\n",
       " 'gex:MSAFHOther',\n",
       " 'gex:PDRiskPesticideExposure',\n",
       " 'gex:PDRiskOccupationalsolventExposure',\n",
       " 'gex:PDRiskDrinksCaffeinatedBeverages',\n",
       " 'gex:PDRiskSmokerorFormerSmoker',\n",
       " 'gex:PDRiskPhysicallyActive',\n",
       " 'gex:MeanStriatumSBR',\n",
       " 'gex:MeanStriatumSBR%ile',\n",
       " 'gex:Caudate/PutamenRatio',\n",
       " 'gex:AssymetryIndex',\n",
       " 'gex:OrthostaticHypotension?',\n",
       " 'gex:BloodyTap?',\n",
       " 'gex:ClinicalCSFDataObtained?',\n",
       " 'gex:CSFProtein',\n",
       " 'gex:CSFGlucose',\n",
       " 'gex:ClinicalCSFRBC',\n",
       " 'gex:ClinicalCSFTotalNucleatedCells',\n",
       " 'gex:ClinicalCSFGranulocytes',\n",
       " 'gex:ClinicalCSFLymphocytes',\n",
       " 'gex:CSFClinicalMonocytes',\n",
       " 'gex:CSFVolume',\n",
       " 'gex:CSFTotalCells',\n",
       " 'gex:CSFCells/uL',\n",
       " 'gex:CSFNotes',\n",
       " 'gex:CSFSAA',\n",
       " 'gex:UPSITRawScore',\n",
       " 'gex:UPSITPercentile',\n",
       " 'gex:MOCAVisuospatial',\n",
       " 'gex:MOCA2Naming',\n",
       " 'gex:MOCAMemoryTrial1',\n",
       " 'gex:MOCAMemoryTrial2',\n",
       " 'gex:MOCAAttention',\n",
       " 'gex:MOCALanguage',\n",
       " 'gex:MOCAAbstraction',\n",
       " 'gex:MOCADelayedRecall',\n",
       " 'gex:MOCAMemoryIndexScore',\n",
       " 'gex:MOCAOrientation',\n",
       " 'gex:MOCATotal',\n",
       " 'gex:ParisArousalTotal',\n",
       " 'gex:ConstipationCriteriaScore',\n",
       " 'gex:BMSQ1RisingfromChair',\n",
       " 'gex:BMSQ2Handwriting',\n",
       " 'gex:BMSQ3VoiceSofter',\n",
       " 'gex:BMSQ4PoorBalance',\n",
       " 'gex:BMSQ5FeetGettingStuck',\n",
       " 'gex:BMSQ6FaceLessExpressive',\n",
       " 'gex:BMSQ7ArmsorLegsShaking',\n",
       " 'gex:BMSQ8Buttons',\n",
       " 'gex:BMSQ9Shuffling',\n",
       " 'gex:BMSQ10MoveMoreSlowlythanOthersofAge',\n",
       " 'gex:BMSQTotal',\n",
       " 'gex:LRScore',\n",
       " 'gex:ProbableProdromalPD?',\n",
       " 'gex:SCOPATotal',\n",
       " 'gex:SCPOPAManualEntry',\n",
       " 'gex:NMSSTotal',\n",
       " 'gex:MDS-UPDRS1.1cognition',\n",
       " 'gex:MDS-UPDRS1.2Hallucinations',\n",
       " 'gex:MDS-UPDRS1.3Mood',\n",
       " 'gex:MDS-UPDRS1.4Anxiety',\n",
       " 'gex:MDS-UPDRS1.5Apathy',\n",
       " 'gex:MDS-UPDRS1.6DOPADysregulation',\n",
       " 'gex:MDS-UPDRS1.7Sleep',\n",
       " 'gex:MDS-UPDRS1.8DaytimeSleepiness',\n",
       " 'gex:MDS-UPDRS1.9Pain',\n",
       " 'gex:UMD-UPDRS1.10Urological',\n",
       " 'gex:MDS-UPDRS1.11Constipation',\n",
       " 'gex:MDS-UPDRS1.12OrthostaticDizziness',\n",
       " 'gex:MDS-UPDRS1.13Fatigue',\n",
       " 'gex:MDS-UPDRS2.1Speech',\n",
       " 'gex:MDS-UPDRS2.2Drooling',\n",
       " 'gex:MDS-UPDRS2.3Swallowing',\n",
       " 'gex:MDS-UPDRS2.4Eating',\n",
       " 'gex:MDS-UPDRS2.5Dressing',\n",
       " 'gex:MDS-UPDRS2.6Hygeine',\n",
       " 'gex:MDS-UPDRS2.7Handwriting',\n",
       " 'gex:MDS-UPDRS2.8Hobbies',\n",
       " 'gex:MDS-UPDRS2.9TurninginBed',\n",
       " 'gex:MDS-UPDRS2.10Tremor',\n",
       " 'gex:MDS-UPDRS2.11RisingfromChair,Car,Bed',\n",
       " 'gex:MDS-UPDRS2.12Walking',\n",
       " 'gex:MDS-UPDRS2.13FreezingofGait',\n",
       " 'gex:MDS-UPDRSTakingMeds',\n",
       " 'gex:MDS-UPDRS3.1Speech',\n",
       " 'gex:MDS-UPDRS3.2Facies',\n",
       " 'gex:MDS-UPDRS3.3NeckRigidity',\n",
       " 'gex:MDS-UPDRS3.3RUERigidity',\n",
       " 'gex:MDS-UPDRS3.3LUERidigity',\n",
       " 'gex:MDS-UPDRS3.3RLERidigity',\n",
       " 'gex:MDS-UPDRS3.3LLERigidty',\n",
       " 'gex:MDS-UPDRS3.4FingerTapsR',\n",
       " 'gex:MDS-UPDRS3.4FingerTapsL',\n",
       " 'gex:MDS-UPDRS3.5GraspR',\n",
       " 'gex:MDS-UPDRS3.5GraspL',\n",
       " 'gex:MDS-UPDRS3.6RSMR',\n",
       " 'gex:MDS-UPDRS3.6RSML',\n",
       " 'gex:MDS-UPDRS3.7ToeTapR',\n",
       " 'gex:MDS-UPDRS3.7ToeTapL',\n",
       " 'gex:MDS-UPDRS3.8LegAgilityR',\n",
       " 'gex:MDS-UPDRS3.8LegAgilityL',\n",
       " 'gex:MDS-UPDRS3.9RisingFromChair',\n",
       " 'gex:MDS-UPDRS3.10Gait',\n",
       " 'gex:MDS-UPDRS3.11FreezingofGait',\n",
       " 'gex:MDS-UPDRS3.12PullTest',\n",
       " 'gex:MDS-UPDRS3.13Posture',\n",
       " 'gex:MDS-UPDRS3.14GlobalSpontenaity',\n",
       " 'gex:MDS-UPDRS3.15PosturalTremorR',\n",
       " 'gex:MDS-UPDRS3.15PosturalTremorL',\n",
       " 'gex:MDS=UPDRS3.16KineticTremorRUE',\n",
       " 'gex:MDS=UPDRS3.16KineticTremorLUE',\n",
       " 'gex:MDS=UPDRS3.17RestTremorAmp.RUE',\n",
       " 'gex:MDS=UPDRS3.17RestTremorAmp.LUE',\n",
       " 'gex:MDS=UPDRS3.17RestTremorAmp.RLE',\n",
       " 'gex:MDS=UPDRS3.17RestTremorAmp.LLE',\n",
       " 'gex:MDS=UPDRS3.17RestTremorAmp.Lips/Jaw',\n",
       " 'gex:MDS-UPDRS3.18ConstancyofRestTremor',\n",
       " 'gex:Hoehn&YahrStage',\n",
       " 'gex:MDS-UPDRSPartITotal',\n",
       " 'gex:MDS-UPDRSPartIITotal',\n",
       " 'gex:MDS-UPDRSPartIIITotal',\n",
       " 'gex:MDS-UPDRSPartI-IIITotal',\n",
       " 'gex:MDS-UPDRSPartIbTotal',\n",
       " 'gex:MDS-UPDRSPartIb+IITotal',\n",
       " 'gex:RBDSQTotal',\n",
       " 'gex:ESSTotal',\n",
       " 'gex:RightCaudateSBR',\n",
       " 'gex:RightAnteriorPutamenSBR',\n",
       " 'gex:RightPosteiorPutamenSBR',\n",
       " 'gex:LeftAnteirorPutamenSBR',\n",
       " 'gex:LeftPosteriorPutamenSBR',\n",
       " 'gex:LeftCaudateSBR',\n",
       " 'gex:MeanStriatumSBR.1',\n",
       " 'gex:MeanStriatumSBRPercentile',\n",
       " 'gex:CaudatePutamenRatio',\n",
       " 'gex:AssymetryIndex.1',\n",
       " 'gex:PhenoconverteasofSept2024r?',\n",
       " 'gex:CSFTNFalpha',\n",
       " 'gex:CSFTNFBeta',\n",
       " 'gex:PlasmaTNFalpha',\n",
       " 'gex:PlasmaTNFBeta',\n",
       " 'gex:doublet_score',\n",
       " 'gex:predicted_doublet',\n",
       " 'gex:n_genes_by_counts',\n",
       " 'gex:log1p_n_genes_by_counts',\n",
       " 'gex:total_counts',\n",
       " 'gex:log1p_total_counts',\n",
       " 'gex:pct_counts_in_top_50_genes',\n",
       " 'gex:pct_counts_in_top_100_genes',\n",
       " 'gex:pct_counts_in_top_200_genes',\n",
       " 'gex:pct_counts_in_top_500_genes',\n",
       " 'gex:total_counts_mt',\n",
       " 'gex:log1p_total_counts_mt',\n",
       " 'gex:pct_counts_mt',\n",
       " 'gex:total_counts_ribo',\n",
       " 'gex:log1p_total_counts_ribo',\n",
       " 'gex:pct_counts_ribo',\n",
       " 'gex:total_counts_hb',\n",
       " 'gex:log1p_total_counts_hb',\n",
       " 'gex:pct_counts_hb',\n",
       " 'gex:n_genes',\n",
       " 'gex:_scvi_batch',\n",
       " 'gex:_scvi_labels',\n",
       " 'gex:predicted_labels_Immune_All_Low',\n",
       " 'gex:over_clustering_Immune_All_Low',\n",
       " 'gex:majority_voting_Immune_All_Low',\n",
       " 'gex:predicted_labels_COVID19_HumanChallenge_Blood',\n",
       " 'gex:over_clustering_COVID19_HumanChallenge_Blood',\n",
       " 'gex:majority_voting_COVID19_HumanChallenge_Blood',\n",
       " 'gex:cluster_L1',\n",
       " 'gex:cluster_L2',\n",
       " 'gex:Hyposmia',\n",
       " 'gex:Diagnosis_RBDstrat',\n",
       " 'gex:Gender',\n",
       " 'gex:Age',\n",
       " 'tcr:cellranger_sample_id',\n",
       " 'tcr:SampleID',\n",
       " 'tcr:SampleType',\n",
       " 'tcr:DonorID',\n",
       " 'tcr:receptor_type',\n",
       " 'tcr:receptor_subtype',\n",
       " 'tcr:chain_pairing',\n",
       " 'bcr:cellranger_sample_id',\n",
       " 'bcr:SampleID',\n",
       " 'bcr:SampleType',\n",
       " 'bcr:DonorID',\n",
       " 'bcr:receptor_type',\n",
       " 'bcr:receptor_subtype',\n",
       " 'bcr:chain_pairing',\n",
       " 'tcr:clone_id',\n",
       " 'tcr:clone_id_size']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(mdata.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gex:EnrollmentCategory</th>\n",
       "      <th>gex:Diagnosis_RBDstrat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-YPD001BLD</th>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-YPD001BLD</th>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCCCTTG-1-YPD001BLD</th>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCCCACG-1-YPD001BLD</th>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACGAGAGT-1-YPD001BLD</th>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCTTCGGTTCCACAA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTTCTAAGCCCGAAA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGAGGCTCTTGCATT-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCTCCTTCGTTACGA-1-YPD094BLD</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGACGTCTCGCAAACT-1-YPD094CSF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1029222 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gex:EnrollmentCategory gex:Diagnosis_RBDstrat\n",
       "AAACCTGAGCTCCTTC-1-YPD001BLD                     HC                     HC\n",
       "AAACCTGAGGACATTA-1-YPD001BLD                     HC                     HC\n",
       "AAACCTGAGGCCCTTG-1-YPD001BLD                     HC                     HC\n",
       "AAACCTGAGTCCCACG-1-YPD001BLD                     HC                     HC\n",
       "AAACCTGCACGAGAGT-1-YPD001BLD                     HC                     HC\n",
       "...                                             ...                    ...\n",
       "GTCTTCGGTTCCACAA-1-YPD094BLD                    NaN                    NaN\n",
       "GTTTCTAAGCCCGAAA-1-YPD094BLD                    NaN                    NaN\n",
       "TCGAGGCTCTTGCATT-1-YPD094BLD                    NaN                    NaN\n",
       "TTCTCCTTCGTTACGA-1-YPD094BLD                    NaN                    NaN\n",
       "GGACGTCTCGCAAACT-1-YPD094CSF                    NaN                    NaN\n",
       "\n",
       "[1029222 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.obs[['gex:EnrollmentCategory', 'gex:Diagnosis_RBDstrat']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell_Name</th>\n",
       "      <th>clone.id</th>\n",
       "      <th>patient</th>\n",
       "      <th>majorCluster</th>\n",
       "      <th>loc</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_rbd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGTCCCACG-1-YPD001BLD</td>\n",
       "      <td>0</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>Tem/Trm cytotoxic T cells</td>\n",
       "      <td>BLD</td>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGCACGAGAGT-1-YPD001BLD</td>\n",
       "      <td>1</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>BLD</td>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGGTGAAGGCT-1-YPD001BLD</td>\n",
       "      <td>2</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>Tem/Trm cytotoxic T cells</td>\n",
       "      <td>BLD</td>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGGGAGATCACGG-1-YPD001BLD</td>\n",
       "      <td>3</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>Regulatory T cells</td>\n",
       "      <td>BLD</td>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGCAAGAGGCT-1-YPD001BLD</td>\n",
       "      <td>4</td>\n",
       "      <td>YPD001</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>BLD</td>\n",
       "      <td>HC</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418057</th>\n",
       "      <td>TTTGTCATCGTTTATC-1-YPD028CSF</td>\n",
       "      <td>204471</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>CSF</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>PD-RBD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418058</th>\n",
       "      <td>TTTGTCATCTCGCATC-1-YPD028CSF</td>\n",
       "      <td>202374</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>CSF</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>PD-RBD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418059</th>\n",
       "      <td>TTTGTCATCTGATTCT-1-YPD028CSF</td>\n",
       "      <td>196485</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>CSF</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>PD-RBD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418060</th>\n",
       "      <td>TTTGTCATCTGGTGTA-1-YPD028CSF</td>\n",
       "      <td>204472</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>CSF</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>PD-RBD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418061</th>\n",
       "      <td>TTTGTCATCTTAACCT-1-YPD028CSF</td>\n",
       "      <td>204473</td>\n",
       "      <td>YPD028</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>CSF</td>\n",
       "      <td>PD-RBD</td>\n",
       "      <td>PD-RBD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>405022 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Cell_Name clone.id patient  \\\n",
       "0       AAACCTGAGTCCCACG-1-YPD001BLD        0  YPD001   \n",
       "1       AAACCTGCACGAGAGT-1-YPD001BLD        1  YPD001   \n",
       "2       AAACCTGGTGAAGGCT-1-YPD001BLD        2  YPD001   \n",
       "3       AAACGGGAGATCACGG-1-YPD001BLD        3  YPD001   \n",
       "4       AAACGGGCAAGAGGCT-1-YPD001BLD        4  YPD001   \n",
       "...                              ...      ...     ...   \n",
       "418057  TTTGTCATCGTTTATC-1-YPD028CSF   204471  YPD028   \n",
       "418058  TTTGTCATCTCGCATC-1-YPD028CSF   202374  YPD028   \n",
       "418059  TTTGTCATCTGATTCT-1-YPD028CSF   196485  YPD028   \n",
       "418060  TTTGTCATCTGGTGTA-1-YPD028CSF   204472  YPD028   \n",
       "418061  TTTGTCATCTTAACCT-1-YPD028CSF   204473  YPD028   \n",
       "\n",
       "                       majorCluster  loc disease disease_rbd  \n",
       "0         Tem/Trm cytotoxic T cells  BLD      HC          HC  \n",
       "1       Tem/Temra cytotoxic T cells  BLD      HC          HC  \n",
       "2         Tem/Trm cytotoxic T cells  BLD      HC          HC  \n",
       "3                Regulatory T cells  BLD      HC          HC  \n",
       "4       Tem/Temra cytotoxic T cells  BLD      HC          HC  \n",
       "...                             ...  ...     ...         ...  \n",
       "418057     Tcm/Naive helper T cells  CSF  PD-RBD      PD-RBD  \n",
       "418058     Tcm/Naive helper T cells  CSF  PD-RBD      PD-RBD  \n",
       "418059     Tcm/Naive helper T cells  CSF  PD-RBD      PD-RBD  \n",
       "418060     Tcm/Naive helper T cells  CSF  PD-RBD      PD-RBD  \n",
       "418061     Tcm/Naive helper T cells  CSF  PD-RBD      PD-RBD  \n",
       "\n",
       "[405022 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_t_crosstissue = ['Tem/Temra cytotoxic T cells',\n",
    " 'Tem/Trm cytotoxic T cells',\n",
    " 'Regulatory T cells',\n",
    " 'Tcm/Naive helper T cells',\n",
    " 'Tem/Effector helper T cells',\n",
    " 'Tcm/Naive cytotoxic T cells',]\n",
    "\n",
    "df_startrac = mdata.obs[['tcr:clone_id', 'gex:DonorID', 'gex:majority_voting_Immune_All_Low', 'gex:SampleType',\n",
    "'gex:EnrollmentCategory', 'gex:Diagnosis_RBDstrat']].dropna().reset_index()\n",
    "df_startrac.columns = ['Cell_Name', 'clone.id', 'patient', 'majorCluster', 'loc', 'disease', 'disease_rbd']\n",
    "# df_startrac['loc'] = 'CSF'\n",
    "df_startrac = df_startrac[df_startrac['majorCluster'].isin(list_t_crosstissue)]\n",
    "df_startrac.to_csv(f'../scanpy/{version}/startrac_input_T_crosstissue.csv')\n",
    "df_startrac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy1.10.2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
